### ISSUE TITLE
[MU4 Issue] interface and score do not scale properly with -D

### ISSUE THREAD

**Issue by MarcSabatella** [musescore/MuseScore#10354]
Un the past, MuseScore has tried to detect the monitor resolution and scale everything - interface and score alike - so that the physical size of all icons and of the score itself is identical across systems.  In practice, this often doesn't work because of the different ways various operating systems attempt to scale things at their end.  So, we provide the "-D" command line option to tell MuseScore the actual resolution.  In MuseScore 3, this works well and provides a consistent look and feel across systems.  In MuseScore 4, it does not work correctly at all.

To reproduce, simply start MuseScore 4 with and without the -D option.  When using it, give it an argument in DPI, eg, "-D 166" to specify that MuseScore should scale itself for a 166 DPI display.

What I see on my 166 DPI display (Linux) is that interface itself doesn't scale at all - it's tiny with or with "-D 166".  The size of the *page* does scale, and so do a few drawn items like note stems, but elements drawn from music fonts do not, leading to just plain terrible result.

Here's a screenshot *without* "-D 166" - as with MuseScore 3 on this system, the interface itself is way too small to be usable, and the score itself renders much too small (required 166% zoom to be actual size):

![image](https://user-images.githubusercontent.com/1799936/150614998-57cdf207-2be3-4c13-ace1-1936af445b04.png)

Here's a screenshot *with* "-D  166".  Interface remains tiny.  Score page is now proper size, but notation looks ridiculous:

![image](https://user-images.githubusercontent.com/1799936/150614939-17a7d075-4e45-4d09-a582-b0e48aa29292.png)

Here's a screenshot comparing MuseScore 3 & 4, with "-D 166", to give an idea of how much smaller everything is:

![image](https://user-images.githubusercontent.com/1799936/150615341-b5650d5f-0bce-4759-bf04-7d390d0792ea.png)

**Comment by cbjeukendrup** [comment]
Another issue that I noticed, is that the score page size does not equal the physical size anymore when zoom is 100% on high-dpi screens (e.g. if you have a score on A4 paper, the size on screen will not match the size of a real sheet of A4 paper). I fixed this a while ago in PR #9112, but apparently this was broken again by the more recent changes to how drawing works.

**Comment by MarcSabatella** [comment]
Fwiw, assuming we can’t do any better at detecting the display resolution than we did in MuseScore 3 (thus eliminating the need for the user overrides and the almost daily support posts about how to fix sizes), I think we should scrap the “-D” option in favor of a simple preference the user can set within the UI.  It would be far simpler for users who don’t have any idea what a command line option is.

For the record, this is kind of a blocking issue for me as I try to start working on documentation.  I need a usable UI that I can take reasonable screenshots of.

**Comment by Eism** [comment]
#10446
Since this decision affects all components of the project, it is obvious that this decision requires extensive testing.
Now, this is an unstable solution for me, there are many problems to be solved.

But at least it solves the main problem.
@MarcSabatella, can you test this solution and write out any bugs you notice?

**Comment by Eism** [comment]
Still, I think standard scaling should be used for documentation because most users don't use scaling(difference in appearance). Or is it due to some technical issue(get the clearest screenshot)?

**Comment by jeetee** [comment]
> I think we should scrap the “-D” option in favor of a simple preference the user can set within the UI. It would be far simpler for users who don’t have any idea what a command line option is.

+5
There are plenty of non-tech savvy users out there that have bumped against this wall in the past; I'd keep the -D for a session override and when trying to figure out a good value *if* the preference needs a restart to take effect. If the preference can be applied instantly, then the cmd line version is likely no longer required.

**Comment by Tantacrul** [comment]
I imagine a general scaling that affects everything would be sensible. I also agree it should be in Preferences under 'Appearance'.

@MarcSabatella the problem we have now is that we want to prioritise features for Alpha. This feels very much like a Beta type of optimisation.

I'm wondering if there is any simple alternative to making the app bigger. Possibly by changing the overall scaling on your machine?

I'm kind of clutching at straws here because I'm aware @Eism has some very pressing tasks in front of him.

Do let us know if his temp solution works.

**Comment by Tantacrul** [comment]
@MarcSabatella - if we can figure out a temporary solution to this, it'd be great. I'd hate for you to be blocked.

**Comment by jeetee** [comment]
> The problem we have now is that we want to prioritise features for Alpha. This feels very much like a Beta type of optimisation.

Agreed

**Comment by MarcSabatella** [comment]
This isn't an "optimization", it's basic functionality.  On a high DPI screen, the buttons are potentially too small to be recognizable or clickable, rendering the UI entirely useless on some displays.  On my system, I can make them out, but my eyes are not what they used to be, and it's extremely difficult for me to work with it as is.

For alpha, simply fixing "-D" seems fine.  Adding a GUI setting can wait, for sure.

**Comment by MarcSabatella** [comment]
@Eism  I tried to access the artifact from the PR but couldn't find it on https://github.com/musescore/MuseScore/runs/5064604225?check_suite_focus=true - can you post a direct link to a Linux build

To be clear: I am not trying to achieve non-standard scaling.  I am simply trying to see normal-sized icons and text.

Right now, the menu text is minuscule, as are all icons - toolbar and palette.  In addition to "-D" which formerly fixed everything, there used to be preferences for icon and palette scaling but the former is gone and the latter doesn't work.  So I'm not finding any workaround here.  Aside from the main main, other UI text *does* at least honor the preference setting, so I can force Home / Score / Publish and the palette *labels* to look correct by setting an artificially high font size.  But that doesn't help with the rest.

**Comment by MarcSabatella** [comment]
If I change the display resolution system-wide, some things scale but others don't, making it somewhat more usable but now the different UI elements are out of proportion with each other and not suitable for screenshots.

I doubt I'm the only user who uses a high DPI display, so I wouldn't assume this is just a special for my benefit.

**Comment by Eism** [comment]
> @Eism I tried to access the artifact from the PR but couldn't find it on https://github.com/musescore/MuseScore/runs/5064604225?check_suite_focus=true - can you post a direct link to a Linux build

@MarcSabatella 
https://github.com/musescore/MuseScore/suites/5177473248/artifacts/157179768

**Comment by Jojo-Schmitz** [comment]
Or https://github.com/musescore/MuseScore/actions/runs/1794250824

**Comment by MarcSabatella** [comment]
Thanks for the links!  I guess mine had the extra "check_suite_focus" added by GitHub and that's why I couldnm't get to the artifacts?

Anyhow, it's a bit better using the artifact, still far from ideal though.  No change without -D, but when using it, the toolbar icons and interface text do scale.  Everything else is still too small though - the main menu (File / Edit / etc), the palette icons, and the score itself.  Also, the height of the toolbar doesn't scale, so the icons are now clipped at top & bottom.

![image](https://user-images.githubusercontent.com/1799936/152567536-c4b37f8c-8183-4592-8d90-25c2fed988c7.png)

But also seems that the calculation is off because I feel like I need to use a value for -D that is less than my actual resolution to get what "feels" right.  On the other hand, since I've never actually seen MU4 at the "correct" size, it's hard for me to judge how it's supposed to look.

For the record, we used to have three very simple objective measurements here:

1) the score at 100% zoom should be actual size - hold up a piece of paper to the screen, it should match
2) the palette icons should be the same size as the actual score elements (except for those palettes that were deliberately downscaled in the individual palette or cell properties)
3) the UI font size specified in Edit / Preferences / General should be literally correctly those font sizes - a text font size of 12pt should literally be 12 points high on the screen

The third was always the hardest to get right, because the OS *really* wants to have a say in font scaling.  So I often resorted to specifying a different font size than I actually wanted so it came out right in the end.

**Comment by MarcSabatella** [comment]
I recently updated my Windows system to Windows 11 and haven't tried fiddling with the various options there, including whatever Windows 11 now provided for high DPI compatibility settings.  I'll do some more experimentation there in a bit to see if maybe I can now at least get that system to display something reasonable.

**Comment by MarcSabatella** [comment]
Situation on Windows is better indeed.  My display resolution on Windows is the as on my Linux system, and in MuseScore 3 it required similar methods to scale correctly.  But MU4 looks mostly good out of the box using the current nightly (not the PR artifact; I haven't tried that).  The main menu looks right, and the toolbar and palette icons look right.  But the score is still too small, and the other text is also way smaller than I expect - palette labels, Home / Score / Publish, etc.  Also the *contents* of the main menu are way too small: e.g., the word "Help" is normal sized, but the "Online handbook", "About...", "About Qt...", etc are tiny:

![Screenshot 2022-02-04 111149](https://user-images.githubusercontent.com/1799936/152581245-7147847a-a086-4a9a-8695-edef656d4266.png)

The good news here is, increasing the font size in preferences fixes that - is the default really meant to be 9pt? - and of course I can zoom in to the score.  So while the result still requires expert-level intervention - not yet suitable for ordinary users with high DPI displays - it's usable by me at least.

**Comment by MarcSabatella** [comment]
In order for menu contents to display the same size as the menu itself, I need to set 12pt as the font size in Preferences.  And this also makes the other interface text look more reasonable to me.  The palette labels are the same size as the menu contents; the Home / Score / Publish is slightly larger.  I guess that's on purpose?

BTW, is it a known issue that spinboxes don't work with the cursor up/down keys?  That's true across the whole program as far as I can tell - Properties window, Preferences dialog, etc.

**Comment by Tantacrul** [comment]
9pt shouldn't be used anywhere. The minimum size we use across the app is 12pt. Something must be up, although I can't say what. Windows and Mac builds are looking fine / standard on various different resolutions for me. 

Just to be sure, can you reset to factory settings?

Obviously, we've a bit more hunting to do. One point though regarding screens with high DPI, they generally just scale to a different %. For example, most Windows laptops these days scale to 125%, 150% and 200% by default so MuseScore really should not be relied upon for that.

However, the idea of just having a few scaling settings for people who want a very large UI (but don't want to use OS scaling) is still sensible and we should do it.

**Comment by MarcSabatella** [comment]
OK, that makes sense.  I’m not at my Wi does computer but will try a factory reset when I get there.  I can confirm on my Linux machine, factory reset does result in 12pt, making the menu co tents exactly as tiny as everything else. :-)

It’s true that *in theory*, scaling should work correctly out of the box based on the OS native handling.  But in practice, that just isn’t so.  That’s why almost every day we get someone posting to the forum or FB or Reddit or wherever, “help, everything is tiny”.    Or occasionally, “help, everything is huge”.  

Of my own systems that I’ve run MuseScore 3 on over the last few years for, the problem exists for 4 out of 5 m, which is *not* a good average.  Everything else looks normal, MuseScore is either tiny or huge.  Tiny on both my Linux systems, tiny on one Windows system, huge on another, just right on the third.

 So unfortunately it’s *not* an uncommon problem at all.  That why we provided the -D option in the first place.   Not to allow users to create custom non-standard scaling (although that’s nice to have), but as a workaround for the very large number of systems that somehow just don’t get the scaling right out of the box.

Maybe the new QML UI will make this happen less often, so sure, we can roll the dice with the alpha and hope that turns out to be true.  But right now I’m only 1 for 3 (unusable on two Linux systems, OK on the one Windows system I’ve tried).

So I don’t have a great feeling about it, but all I have to base that on my own experience.  Would be interesting to maybe put out a call on the forum or wherever to get people to install a specific nightly build and just report back if the size seems OK.

**Comment by Tantacrul** [comment]
I’m pretty sure we can solve the majority of them. We should just start logging each instance as a bug (and we’ll prioritise each one).

We’ll have the fallback too, just in case.

**Comment by MarcSabatella** [comment]
Late-breaking good news - on one of my Linux systems at least (still need to test the other later), MU4 responds to the QT_AUTO_SCREEN_SCALE_FACTOR environment variable - setting it to 1 enables auto-sizing and causes the GUI to be sized correctly.  The score itself is still too small, but that's a completely usable workaround otherwise.  Weird, in that I thought this environment had been obsoleted, but I guess not - at least, not as long as we stick with Qt 5.

There is also QT_SCALE_FACTOR that you can set to a specific real value like 1.66 or 3.0 or whatever.  This works also but somehow not quite correctly for me, the scores actually gets too big too fast compared to the GUI.  So, with the correct scaling factor to get normal-sized text and icons, my score is twice as big big as it should be.  But, I like this option anyhow because it lets me choose a compromise value for the scaling on my smaller-screened machine where I don't actually want *full* size GUI, just bigger than the microscopic default.

So anyhow, I'm still curious to see how this plays out in the wild, but anyhow, consider me unblocked.

**Comment by MarcSabatella** [comment]
... and now I see why I believe QT_AUTO_SCREEN_SCALE_FACTOR didn't work anymore (other than the fact I thought I had read it was being obsoleted) - because it *doesn't* work on my main system.  Only on my smaller tablet-based system, which is what I was testing earlier today.  However, QT_SCALE_FACTOR *does* work on my main system.  I'm still not really sure about the GUI vs score scaling.  I wonder if there is an objective way to say how big things are *supposed* to be, the way there always was in MuseScore 3?  At this point I don't really know what I should be expecting.  Buth here's my attempt at measuring objectively:

- Through trial and error, I determined that a QT_SCALE_FACTOR value of 1.38 yields a score that is the correct size (as measured by physically holding up a sheet of paper to my monitor for comparison).  The palette icons are to a rough approximation the same size as the score, but are actually around 10% too small, as you can see by comparing the size of the staff in the key signature palette to the staff in the score itself:
![image](https://user-images.githubusercontent.com/1799936/152657368-db0dd460-cdc0-46e3-acda-c7d0b9100158.png)

- If I set the text font in my score to the same as my UI (Fira Sans, 12 pt), the UI is very clearly still quite a bit smaller, although it's certainly comfortable compared to what I'm used to after seeing the tiny fonts I've been seeing until now:
![image](https://user-images.githubusercontent.com/1799936/152657551-6709ec53-0f4a-48d7-b510-9ed418ad7186.png)

- Assuming the 12pt in my score really is literally 12 points tall, further comparison tells me the UI font is actually displaying at less than 8 points (8 is now the minimum value I can select for text size?).  So, question, if the preference is set to 12pt, does that mean we really are literally expecting 12pt text in the UI?  That actually seems like it would be quite big.  Or maybe it means, "some" text is 12pt, but other text is scaled down from there?  In particular, since it's the easiest to measure - how tall should the text be on the score tab saying "Untitled"?

- I know my display is actually 166 DPI.  Which isn't actually all that high of course, but the baseline is usually taken to be 100 DPI, so I normally use a scaling factor of 1.66 for my "-D" option and it's always yielded the correct results for score & palette size before.  But in MU4, if I set QT_SCALE_FACTOR to 1.66, the score is definitely too big.  Setting the zoom to 83% makes it the right size.  So, now I can compare size of text in my score to the UI, and I find that the actual effective size is 9pt:
![image](https://user-images.githubusercontent.com/1799936/152657967-ef79c2f2-528f-4194-a1ee-4d331afa2645.png)

- As it happens, 9pt is exactly what I always have my UI font set to in MuseScore 3. but it was actually very clearly *less* than 9pt in reality, because of how the scaling was working there.  In MuseScore 3, with my UI font set to 9pt, it actually matched 7pt font in the score:
![image](https://user-images.githubusercontent.com/1799936/152658290-83ab5593-6d94-4ccf-995e-43d5cc2d1816.png)

I'm not sure what this all adds up to.  Again, I have some *usable*, but I'm still at a loss to say what's *correct*.

**Comment by MarcSabatella** [comment]
OK, here's a simpler comparison, MuseScore 3 next to MuseScore 4.  No idea why this should be the case, but it seems that with MuseScore 3 set to 9pt UI font, and MuseScore 4 set to 12pt with QT_SCALE_FACTOR set to 1.38, it's a more or less exact match:

![image](https://user-images.githubusercontent.com/1799936/152658995-31840ef5-d1ce-4c7c-b014-0b617b6686ec.png)

The actual size of the UI font - as measured by comparing to text in the score itself - is more like 7pt in both cases, but it's what I'm used to and consistent with other applications, so I assume it's OK.  The score is the correct size.

So, I *think* I'm seeing what I'm supposed to be seeing.  Do the relative sizes shown in this image look correct to people?

**Comment by Tantacrul** [comment]
MS4 looks about right to me in that image, yes. 

Incidentally, in places the default scaling of MS3 is much larger than MS4 on my machines (mainly the UI icons for the note durations). Other things, like palettes are roughly the same.

**Comment by Tantacrul** [comment]
I believe this issue is resolved for now?

**Comment by MarcSabatella** [comment]
For now, yes,  -D does not work as it did before - it used to scale both GUI and the score itself, whereas now it scales the score only - but at least it scales the score *correctly*.  And QT_SCALE_FACTOR also works differently than it did even a few months ago - it now scales the GUI but not the score.  But, that's fine, we now have independent control of these.  So we could simply say, this is how it is now designed to work.

Something is still funny about the actual values, though - they are being multiplied by something, as I still need to give a lower value for -D and for QT_SCALE_FACTOR than a direct calculation tells me I should.  But realistically, every system is different here, so it's not unexpected to need some trial and error.

**Comment by MarcSabatella** [comment]
OK, I realize this may or may not end up being something that affects a lot of other people, but for the record, I just realized that while "-D" *does* work to scale the score when first invoking MuseScore, it does not seem to be for any additional windows that get created when working with multiple scores.  I *think* the issue is simply that command-line arguments that were originally passed to the program, do not also get passed along to additional processes that get started.

I can workaround this too, of course, so no need to reopen this issue at the moment.  But I wanted to log this for if/when we do decide to actually try to deal with this.  I'm updating the original description as well so it will be accurate.

**Comment by Eism** [comment]
@MarcSabatella 
Can you check this PR https://github.com/musescore/MuseScore/pull/10285. Maybe the solution will help you
I enabled fractional scaling. But I didn’t merge it to master, because on Windows score quickly blurry
